### PR TITLE
feature(zc): add plan diffuser

### PR DESCRIPTION
**Pull request by Super1ce**
## Description
add plan diffuser, add normalizer, add dataset

## Check List

- [ ] merge the latest version source branch/repo, and resolve all the conflicts
- [ ] pass style check
- [ ] pass all the tests

### CODE REVIEW

File: ding/policy/command_mode_policy_instance.py [hunk 1]
```diff
@@ -398,6 +399,10 @@ class BCQCommandModelPolicy(BCQPolicy, DummyCommandModePolicy):
 class EDACCommandModelPolicy(EDACPolicy, DummyCommandModePolicy):
     pass
 
+@POLICY_REGISTRY.register('pd_command')
+class EDACCommandModelPolicy(PDPolicy, DummyCommandModePolicy):
```
Comment (PaParaZz1) [inline ding/policy/command_mode_policy_instance.py:403]: class name typo

File: dizoo/d4rl/entry/d4rl_dd_main.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
+from ding.entry import serial_pipeline_offline
```
Comment (PaParaZz1) [inline dizoo/d4rl/entry/d4rl_dd_main.py:1]: remove dd related codes

File: dizoo/d4rl/entry/d4rl_pd_main.py [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+from ding.config import read_config
+from pathlib import Path
+
+
+def train(args):
+    # launch from anywhere
+    config = Path(__file__).absolute().parent.parent / 'config' / args.config 
+    config = read_config(str(config))
+    config[0].exp_name = config[0].exp_name.replace('0', str(args.seed))
+    serial_pipeline_offline(config, seed=args.seed)
```
Comment (PaParaZz1) [inline dizoo/d4rl/entry/d4rl_pd_main.py:11]: set max_train_iter

File: ding/policy/plan_diffuser.py [hunk 4]
```diff
@@ -0,0 +1,339 @@
...
+from ding.rl_utils import v_1step_td_data, v_1step_td_error, get_train_sample, \
+    qrdqn_nstep_td_data, qrdqn_nstep_td_error, get_nstep_return_data
+from ding.policy import Policy
+from ding.model import model_wrap
+from ding.utils import POLICY_REGISTRY, DatasetNormalizer
+from ding.utils.data import default_collate, default_decollate
+from .common_utils import default_preprocess_learn
+
+@POLICY_REGISTRY.register('pd')
+class PDPolicy(Policy):
```
Comment (PaParaZz1) [inline ding/policy/plan_diffuser.py:19]: add paper link

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: ding/model/template/diffusion.py [hunk 5]
```diff
@@ -0,0 +1,617 @@
...
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from ding.utils import list_split, MODEL_REGISTRY, squeeze, SequenceType
+from ding.torch_utils.network.diffusion import extract, cosine_beta_schedule, apply_conditioning, \
+    TemporalUnet, TemporalValue, ARInvModel
+
+Sample = namedtuple('Sample', 'trajectories values chains')
+
+def default_sample_fn(self, x, cond, t):
```
Comment (Mossforest) [inline ding/model/template/diffusion.py:14]: indent & arg 'self' -> should be model?

File: ding/model/template/diffusion.py [hunk 6]
```diff
@@ -0,0 +1,617 @@
...
+        alphas_cumprod = torch.cumprod(alphas, axis=0)
+        alphas_cumprod_prev = torch.cat([torch.ones(1), alphas_cumprod[:-1]])
+        self.n_timesteps = int(n_timesteps)
+
+        self.register_buffer('betas', betas)
+        self.register_buffer('alphas_cumprod', alphas_cumprod)
+        self.register_buffer('alphas_cumprod_prev', alphas_cumprod_prev)
+
+        # calculations for diffusion q(x_t | x_{t-1}) and others
+        self.register_buffer('sqrt_alphas_cumprod', torch.sqrt(alphas_cumprod))
```
Comment (Cloud-Pku) [inline ding/model/template/diffusion.py:112]: where is the definition of the function "register_buffer"?

File: ding/model/template/diffusion.py [hunk 7]
```diff
@@ -0,0 +1,617 @@
+from typing import Union, List, Dict
```
Comment (Cloud-Pku) [inline ding/model/template/diffusion.py:1]: I think you can add some unit tests to make the code easier to understand.
And it's convenient to check the rightness for each module.

File: ding/model/template/diffusion.py [hunk 8]
```diff
@@ -0,0 +1,617 @@
...
+        '''
+        if self.predict_epsilon:
+            return (
+                extract(self.sqrt_recip_alphas_cumprod, t, x_t.shape) * x_t -
+                extract(self.sqrt_recipm1_alphas_cumprod, t, x_t.shape) * noise
+            )
+        else:
+            return noise
+
+    def q_posterior(self, x_start, x_t, t):
```
Comment (Cloud-Pku) [inline ding/model/template/diffusion.py:172]: add comments and use cases

File: ding/torch_utils/network/diffusion.py [hunk 9]
```diff
@@ -0,0 +1,646 @@
...
+            condition_dropout: float = 0.1,
+            calc_energy: bool = False,
+            kernel_size: int = 5,
+            attention: bool = False,
+    ) -> None:
+        super().__init__()
+        dims = [transition_dim, *map(lambda m: dim * m, dim_mults)]
+        in_out = list(zip(dims[:-1], dims[1:]))
+        
+        if calc_energy:
```
Comment (PaParaZz1) [inline ding/torch_utils/network/diffusion.py:239]: why name `calc_energy` here

File: ding/torch_utils/network/diffusion.py [hunk 10]
```diff
@@ -0,0 +1,646 @@
...
+        x = self.mid_down1(x)
+
+        x = self.mid_block2(x, t)
+        x = self.mid_down2(x)
+
+        x = x.view(len(x), -1)
+        out = self.final_block(torch.cat([x, t], dim=-1))
+        return out
+
+class MLPnet(nn.Module):
```
Comment (PaParaZz1) [inline ding/torch_utils/network/diffusion.py:489]: where is this module used?

File: ding/torch_utils/network/diffusion.py [hunk 11]
```diff
@@ -0,0 +1,646 @@
...
+        out = self.mlp(inputs)
+
+        if self.calc_energy:
+            energy = ((out - x) ** 2).mean()
+            grad = torch.autograd.grad(outputs=energy, inputs=x, create_graph=True)
+            return grad[0]
+        else:
+            return out
+        
+class ARInvModel(nn.Module):
```
Comment (PaParaZz1) [inline ding/torch_utils/network/diffusion.py:567]: remove this dd model

File: ding/policy/plan_diffuser.py [hunk 12]
```diff
@@ -0,0 +1,345 @@
...
+    def _init_collect(self) -> None:
+        pass
+
+    def _forward_collect(self, data: dict, **kwargs) -> dict:
+        pass
+
+    def _process_transition(self, obs: Any, model_output: dict, timestep: namedtuple) -> dict:
+        pass
+
+    def _get_train_sample(self, data: list) -> Union[None, List[Any]]:
```
Comment (PaParaZz1) [inline ding/policy/plan_diffuser.py:335]: why not also set this method to pass?

File: ding/policy/plan_diffuser.py [hunk 13]
```diff
@@ -0,0 +1,345 @@
...
+            use_priority_IS_weight=self._cfg.priority_IS_weight,
+            ignore_done=self._cfg.learn.ignore_done,
+            use_nstep=False
+        )
+        
+        conds = {}
+        val = data['condition_val']
+        id = data['condition_id']
+        for i in range(len(val)):
+            if id[i].item() in conds:
```
Comment (PaParaZz1) [inline ding/policy/plan_diffuser.py:222]: you can first use `id = id.tolist()`, then directly use this list rather than the `item` operation.

File: ding/policy/plan_diffuser.py [hunk 14]
```diff
@@ -0,0 +1,345 @@
...
+                conds[id[i].item()].append(val[i])
+            else:
```
Comment (PaParaZz1) [inline ding/policy/plan_diffuser.py:224]: to refine this if-else branch, you can use `defaultdict(list)` here

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

**Comment by PaParaZz1** [comment]
This PR have been manually merged.

File: dizoo/d4rl/entry/d4rl_pd_main.py [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+
+
+if __name__ == "__main__":
+    import argparse
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--seed', '-s', type=int, default=10)
+    parser.add_argument('--config', '-c', type=str, default='hopper_expert_cql_config.py')
```
Comment (puyuan1996) [inline dizoo/d4rl/entry/d4rl_pd_main.py:19]: 默认config应该改成plan diffuser相关的
